### PR TITLE
add fast bandpass filter

### PR DESCRIPTION
**Pull request by timonmerk**
Based on #26 I want to push a faster bandpass filter estimation than the [mne.filter.filter_data](https://mne.tools/stable/generated/mne.filter.filter_data.html).

I also wrote a notebook where I compare the two on my private repo: 
https://github.com/neuromodulation/py_neuromodulation/blob/main/examples/speedtests/speedtest_filter_vs_MNE.ipynb

There you can see that the initialized rt_filter.py estimation took ~6 ms while the mne.filter.filter_data took ~31 ms. Both is based on a sampling frequency of 500 Hz and 6 channels. 

@teonbrooks  Do you think it might be a good idea to upload such an example as well maybe in a folder called "real time tests"?

### CODE REVIEW

File: mne_realtime/rt_filter.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+"""
+Created on Fri Mar 19 13:30:38 2021
+
+@author: Timon Merk
+"""
+
+import numpy as np 
+import mne 
+
+def calc_band_filters(f_ranges, sample_rate, filter_len="1000ms", l_trans_bandwidth=4, h_trans_bandwidth=4):
```
[the remark this answers is not in the document]
Comment (teonbrooks) [reply mne_realtime/rt_filter.py:11]: for `sample_rate`, by convention, we call this `sfreq`

File: mne_realtime/rt_filter.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    filter_list = []
+    for a, f_range in enumerate(f_ranges):
+        h = mne.filter.create_filter(None, sample_rate, l_freq=f_range[0], h_freq=f_range[1], fir_design='firwin',
+                                        l_trans_bandwidth=l_trans_bandwidth, h_trans_bandwidth=h_trans_bandwidth,
+                                        filter_length=filter_len)
+        filter_list.append(h)
+    filter_bank = np.vstack(filter_list)
+    return filter_bank
+
+def apply_filter(dat_, filter_bank, fs):
```
Comment (teonbrooks) [inline mne_realtime/rt_filter.py:41]: I would use data to be clear. usually names with underscores imply that they have been modified (usually in an object like in sklearn)

Comment (teonbrooks) [reply mne_realtime/rt_filter.py:41]: fs is missing in your parameter list

File: mne_realtime/rt_filter.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Mar 19 13:30:38 2021
```
Comment (teonbrooks) [inline mne_realtime/rt_filter.py:3]: as a historical practice, we typically have a commented out block that specify:
`# Author: Author_name <email_address>`

File: mne_realtime/rt_filter.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    filter_list = []
+    for a, f_range in enumerate(f_ranges):
+        h = mne.filter.create_filter(None, sample_rate, l_freq=f_range[0], h_freq=f_range[1], fir_design='firwin',
+                                        l_trans_bandwidth=l_trans_bandwidth, h_trans_bandwidth=h_trans_bandwidth,
+                                        filter_length=filter_len)
+        filter_list.append(h)
+    filter_bank = np.vstack(filter_list)
+    return filter_bank
+
```
Comment (teonbrooks) [inline mne_realtime/rt_filter.py:40]: two lines between functions

File: mne_realtime/rt_filter.py (hunk 2 above)
Comment (teonbrooks) [inline mne_realtime/rt_filter.py:41]: we typically arrange functions in a file in alphabetic order for ease of finding

**Comment by larsoner** [comment]
> There you can see that the initialized rt_filter.py estimation took ~6 ms while the mne.filter.filter_data took ~31 ms. Both is based on a sampling frequency of 500 Hz and 6 channels.


I would much rather speed up `mne.filter.filter_data` for this case rather than add a new function. It's cleaner conceptually and also will benefit more people (anyone using `filter_data` on similar data).

In this case I'm guessing that a direct call to `scipy.signal.lfilter` to do the filtering would be faster than `mne.filter.filter_data`, but it's worth figuring out the conditions under which each method is fastest based on FIR filter length and using `lfilter` or our internal overlap-add method when appropriate.

**Comment by timonmerk** [comment]
@teonbrooks  Thanks a lot for the code review! I will adapt it. 
@larsoner That's a good point, in my implementation it get's only faster since I am precalculating in `calc_band_filters()` the filters using `mne.filter.create_filter`. 

Then I only "apply" them in a np convolution in `def apply_filter()`. Probably this is not an optimal implementation for mne-python :)

**Comment by larsoner** [comment]
> That's a good point, in my implementation it get's only faster since I am precalculating in calc_band_filters() the filters using mne.filter.create_filter.

Ahh, and we don't provide a way to pass the FIR filter directly. We do this with IIR but not FIR...

We could in theory provide some way to do this in MNE. For maximum efficiency you could even compute the FFT of the filter once, and use it with the signal. We'd have to figure out an API for it, though.

**Comment by timonmerk** [comment]
@larsoner Do you mean the iir_params keyword in mne.filter.filter_data? Also I am not sure about the FFT approach. This would also require a frequency transformed version of the data itself? 

Sorry if I don't really get this. For my intuition the fast approach would be simply to do the filtering in the time domain via a convolution with the precalculated filterbank.

File: mne_realtime/rt_filter.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (timonmerk) [reply mne_realtime/rt_filter.py:11]: [mne-filter-filter_data](https://mne.tools/stable/generated/mne.filter.filter_data.html) actually also uses for it's filter_length parameter as a string: 

> filter_length str | int
>     Length of the FIR filter to use (if applicable):
>         ‘auto’ (default): The filter length is chosen based on the size of the transition regions (6.6 times the reciprocal of the shortest transition band for fir_window=’hamming’ and fir_design=”firwin2”, and half that for “firwin”).
>         str: A human-readable time in units of “s” or “ms” (e.g., “10s” or “5500ms”) will be converted to that number of samples if phase="zero", or the shortest power-of-two length at least that duration for phase="zero-double".
>         int: Specified length in samples. For fir_design=”firwin”, this should not be used.

 Apart from that the direct sample rate can be supplied as an int. Maybe the string gives more ease of use this the unit doesn't need to be fixed.

File: mne_realtime/rt_filter.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    Thus the filters can be sequentially used for band power estimation.
+    Parameters
+    ----------
+    f_ranges : list of lists
+        frequency ranges.
+    sample_rate : float
+        sampling frequency.
+    filter_len : str, optional
+        length of the filter. Human readable (e.g."1000ms" or "1s"). Default is "1000ms"
+    l_trans_bandwidth : int/float, optional
```
[the remark this answers is not in the document]
Comment (timonmerk) [reply mne_realtime/rt_filter.py:23]: Alright, I suppose you mean for the `l_trans_bandwidth `and `h_trans_bandwidth `?

File: mne_realtime/rt_filter.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Length of the lower transition band. The default is 4.
+    h_trans_bandwidth : int/float, optional
+        Length of the higher transition band. The default is 4.
+    Returns
+    -------
+    filter_bank : array
```
[the remark this answers is not in the document]
Comment (timonmerk) [reply mne_realtime/rt_filter.py:29]: Yes, ndarray seems to be the nomenclature in mne to specify an array (https://mne.tools/stable/generated/mne.filter.filter_data.html)

File: mne_realtime/rt_filter.py (hunk 2 above)
Comment (timonmerk) [reply mne_realtime/rt_filter.py:41]: Thanks!

Comment (timonmerk) [reply mne_realtime/rt_filter.py:41]: Ohh ok, that good to know. The files are here ordered systematically, so the calc_band_filters comes before apply_filter. But will swap it

**Comment by timonmerk** [comment]
Alright, so I adapted some changes, simply because of the nice comments of @teonbrooks :) 
I am not sure though if this should be included in mne real-time now because of the upper discussion. If so I think a module in test might make sense? 
Similar to my private speed test here https://github.com/neuromodulation/py_neuromodulation/blob/main/examples/speedtests/speedtest_filter_vs_MNE.ipynb

**Comment by larsoner** [comment]
> I am not sure though if this should be included in mne real-time now because of the upper discussion. 

I think for now just add a comment along the lines that "someday MNE-Python could add support for precomputed FIR filters like it does for IIR" and in the meantime a PR just using `np.convolve` directly is fine. You might also want to note that fftconvolve or overlap-add (MNE-Python uses the latter) will be faster in some cases so if this ends up being a bottleneck there are options for optimization.

Then this PR can just focus on getting the API right and getting it to work, and optimization / refactoring of MNE-Python can be done later if someone gets motivated to do it by performance concerns.

**Comment by teonbrooks** [comment]
hey @timonmerk, I will have next week to work on the project. sorry i've run away for a while. If you want to make some final adjustment, I can then merge these and we can iterate on a faster cadence.

**Comment by teonbrooks** [comment]
sorry for the delay @timonmerk. thanks!

**Pull request merged**